helper/keysutil/encrypted_key_storage.go

```diff
@@ -0,0 +1,267 @@
+	// a new enough convergent version. This is a requirement for this storage
+	// implementation.
+	ErrPolicyConvergentVersion = errors.New("key policy must use convergent version > 2")
+
+	// ErrNilStorage is returned if the provided storage is nil.
+	ErrNilStorage = errors.New("nil storage provided")
+)
+
+// EncryptedKeyStorageConfig is used to configure an EncryptedKeyStorage object.
+type EncryptedKeyStorageConfig struct {
+	// Storage is the underlying storage to wrap requests to.
+	Storage logical.Storage
+
+	// Policy is the key policy to use to encrypt the key paths.
+	Policy *Policy
+
+	// Prefix is the storage prefix for this instance of the EncryptedKeyStorage
+	// object. This is stored in plaintext. If not set the DefaultPrefix will be
+	// used.
+	Prefix string
+
+	// CacheSize is the number of elements to cache. If not set the
+	// DetaultCacheSize will be used.
+	CacheSize int
+}
+
+// NewEncryptedKeyStorage takes an EncryptedKeyStorageConfig and returns a new
+// EncryptedKeyStorage object.
+func NewEncryptedKeyStorage(config EncryptedKeyStorageConfig) (*EncryptedKeyStorage, error) {
+	if !config.Policy.Derived {
+		return nil, ErrPolicyDerivedKeys
+	}
+
+	if !config.Policy.ConvergentEncryption {
+		return nil, ErrPolicyConvergentEncryption
+	}
+
+	if config.Policy.ConvergentVersion < 2 {
+		return nil, ErrPolicyConvergentVersion
+	}
+
+	if config.Storage == nil {
+		return nil, ErrNilStorage
+	}
+
+	if config.Prefix == "" {
+		config.Prefix = DefaultPrefix
+	}
+
+	if !strings.HasSuffix(config.Prefix, "/") {
+		config.Prefix += "/"
+	}
+
+	size := config.CacheSize
+	if size <= 0 {
+		size = DefaultCacheSize
+	}
+
+	cache, err := lru.New2Q(size)
+	if err != nil {
+		return nil, err
+	}
+
+	return &EncryptedKeyStorage{
+		policy: config.Policy,
+		s:      config.Storage,
+		prefix: config.Prefix,
+		lru:    cache,
+	}, nil
+}
+
+// EncryptedKeyStorage implements the logical.Storage interface and ensures the
+// storage paths are encrypted in the underlying storage.
+type EncryptedKeyStorage struct {
+	policy *Policy
+	s      logical.Storage
+	lru    *lru.TwoQueueCache
+
+	prefix string
+}
+
+// List implements the logical.Storage List method, and decrypts all the items
+// in a path prefix. This can only operate on full folder structures so the
+// prefix should end in a "/".
+func (s *EncryptedKeyStorage) List(ctx context.Context, prefix string) ([]string, error) {
+	encPrefix, err := s.encryptPath(prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	keys, err := s.s.List(ctx, encPrefix+"/")
+	if err != nil {
+		return keys, err
+	}
+
+	decryptedKeys := make([]string, len(keys))
+
+	// The context for the decryption operations will be the object's prefix
+	// joined with the provided prefix. Join cleans the path ensuring there
+	// isn't a trailing "/".
+	context := []byte(paths.Join(s.prefix, prefix))
+
+	for i, k := range keys {
+		raw, ok := s.lru.Get(k)
```

Comment:
`List` returns just keys, not full paths. As a result, I believe this lookup is problematic; IIUC a list on `a/` returning `[y, z]` and a list on `b/` returning `[y, z]`. The cache will then store values y and z, with values that are dependent on the path, and lookup will then sometimes return the right thing and sometimes not.

Reply:
I think this is okay as is because the values returned by a list on `a/` will be encrypted with a key that is derived by a context that includes `a/`. So a list on `a/` will return `[e(context, y), e(context, z)]` and a list on `b/` will return `[e(context, y), e(context, z)]` where context is unique to each list request.

Reply:
Ah, you're right. It's probably fine.